File: app/src/main/java/com/fanwe/www/http/DownloadActivity.java
package com.fanwe.www.http;

import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.View;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.fanwe.lib.http.Request;
import com.fanwe.lib.http.RequestManager;
import com.fanwe.lib.http.callback.FileRequestCallback;
import com.fanwe.lib.http.utils.TransmitParam;

import java.io.File;

public class DownloadActivity extends AppCompatActivity
{
    public static final String TAG = "DownloadActivity";

    public static final String URL_FILE = "http://110.81.153.150/mmgr.myapp.com/msoft/sec/secure/GodDresser/1/2/3/102027/tencentmobilemanager_20170911185709_7.3.0_android_build4155_102027.apk?mkey=59db0a2773ba1ece&f=d273&c=0&p=.apk";

    private ProgressBar mProgressBar;
    private TextView mTvProgress, mTvSpeed;

    @Override
    protected void onCreate(Bundle savedInstanceState)
    {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_download);
        mProgressBar = (ProgressBar) findViewById(R.id.progressBar);
        mTvProgress = (TextView) findViewById(R.id.tv_progress);
        mTvSpeed = (TextView) findViewById(R.id.tv_speed);
    }

    public void onClickRequest(View view)
    {
        File file = new File(getExternalCacheDir(), "download.apk");
        Request.get(URL_FILE).setTag(this).execute(new FileRequestCallback(file)
        {
            @Override
            protected void onProgressDownload(TransmitParam param)
            {
                mProgressBar.setProgress(param.getProgress());
                mTvProgress.setText(param.getProgress() + "%");
                mTvSpeed.setText(param.getSpeedKBps() + "KB/秒");
                if (param.isFinish())
                {
                    mTvSpeed.setText("");
                    Log.i(TAG, "download finish");
                }
            }

            @Override
            public void onSuccess()
            {

            }

            @Override
            public void onError(Exception e)
            {
                super.onError(e);
                Log.i(TAG, "download error:" + e);
            }

            @Override
            public void onCancel()
            {
                super.onCancel();
                mTvSpeed.setText("");
                Log.i(TAG, "download cancelled");
            }
        });
    }

    public void onClickCancelRequest(View view)
    {
        RequestManager.getInstance().cancelTag(this);
    }

    @Override
    protected void onDestroy()
    {
        super.onDestroy();
        RequestManager.getInstance().cancelTag(this);
    }
}

File: app/src/main/java/com/sd/example/http/activity/DownloadActivity.kt
package com.sd.example.http.activity

import android.os.Bundle
import android.util.Log
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import com.sd.lib.http.Request
import com.sd.lib.http.RequestManager
import com.sd.lib.http.callback.FileRequestCallback
import com.sd.lib.http.impl.GetRequest
import com.sd.lib.http.utils.TransmitParam
import com.sd.example.http.databinding.ActivityDownloadBinding
import java.io.File

class DownloadActivity : AppCompatActivity(), View.OnClickListener {
    private val TAG = DownloadActivity::class.java.simpleName
    private val URL_FILE = "https://qd.myapp.com/myapp/qqteam/AndroidQQ/mobileqq_android.apk"

    private lateinit var _binding: ActivityDownloadBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        _binding = ActivityDownloadBinding.inflate(layoutInflater)
        setContentView(_binding.root)
    }

    override fun onClick(v: View) {
        when (v) {
            _binding.btnStart -> startDownload()
            _binding.btnStop -> stopDownload()
        }
    }

    private fun startDownload() {
        val request: Request = GetRequest().apply {
            this.baseUrl = URL_FILE
            this.tag = TAG
        }

        val file = File(externalCacheDir, "download.apk")
        request.execute(object : FileRequestCallback(file) {

            override fun onStart() {
                super.onStart()
                Log.i(TAG, "onStart")
            }

            override fun onProgressDownload(param: TransmitParam) {
                _binding.progressBar.progress = param.progress
                _binding.tvProgress.text = "${param.progress}%"
                _binding.tvSpeed.text = "${param.speedKBps}KB/秒"
            }

            override fun onSuccess() {
                Log.i(TAG, "onSuccess")
            }

            override fun onError(e: Exception) {
                super.onError(e)
                Log.i(TAG, "onError:$e")
            }

            override fun onCancel() {
                super.onCancel()
                _binding.progressBar.progress = 0
                _binding.tvProgress.text = ""
                _binding.tvSpeed.text = ""
                Log.i(TAG, "onCancel")
            }
        })
    }

    private fun stopDownload() {
        RequestManager.cancelTag(TAG)
    }

    override fun onDestroy() {
        super.onDestroy()
        stopDownload()
    }
}